django/utils/regex_helper.py

```diff
@@ -128,7 +129,13 @@ def normalize(pattern):
                     walk_to_end(ch, pattern_iter)
                 else:
                     ch, escaped = next(pattern_iter)
-                    if ch in "iLmsu#!=<":
+                    deprecated_char = ch in "iLmsu#"
+                    if deprecated_char or ch in "!=<":
+                        if deprecated_char:
+                            warnings.warn(
+                                "Using (?%s) in url() patterns is deprecated." % ch,
+                                RemovedInDjango21Warning
+                            )
                         # All of these are ignorable. Walk to the end of the
                         # group.
                         walk_to_end(ch, pattern_iter)
```

Comment:
Perhaps it would be clearer to repeat this line:

```python
                    if ch in "iLmsu#":
                        warnings.warn(
                            "Using (?%s) in url() patterns is deprecated." % ch,
                            RemovedInDjango21Warning
                        )
                        # All of these are ignorable. Walk to the end of the
                        # group.
                        walk_to_end(ch, pattern_iter)
                    elif ch in "!=<":
                        # All of these are ignorable. Walk to the end of the
                        # group.
                        walk_to_end(ch, pattern_iter)
```